How to expose Option<String> to Roto?

I'm playing with Roto, and am trying to figure out how to expose strings (ideally, `Option<String>`s) to Roto, and am coming up short. This is going to be more of a question, than an issue, I think, please bear with me.

Ideally, I'd like to have a struct like this:

```rust
struct Request {
  pub headers: HeaderMap,
  pub path: Option<String>,
}
```

And on the Roto side, I'd write something like these:

```roto
# matching a path
match request.path() {
  Some(path) -> accept "yay",
  None -> reject "boo",
}

# matching a header
match request.header("user-agent") {
  Some(agent) -> accept "got one",
  None -> reject "no :(",
}
```

My first obstacle was that I couldn't figure out how to get an `Option<String>` into Roto, so I took a step back, and tried to play with a string, and came up with the following rust code:

```rust
use roto::{roto_method, src, Runtime, Val, Verdict};
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct Request {
    pub path: String,
}

fn main() {
    let s = src!(
        r#"
        filtermap main(request: Request) {
          if request.path() == "/" { accept } else { reject }
        }
        "#
    );
    let mut runtime = Runtime::new();
    runtime.register_clone_type::<Request>("Request").unwrap();

    #[roto_method(runtime, Request)]
    fn path(request: &Request) -> Arc<str> {
        request.path.clone().into()
    }

    let mut compiled = s.compile(runtime).unwrap();
    let func = compiled
        .get_function::<(), (Val<Request>,), Verdict<(), ()>>("main")
        .unwrap();

    let req = Request {
        path: "/".to_string(),
    };
    let res = func.call(&mut (), Val(req));

    println!("{res:#?}");
}
```

I thought this would work - it compiled! But running it yields an error:

```
❯ RUST_BACKTRACE=1 cargo run
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/roto-demo`
thread 'main' panicked at /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/codegen/mod.rs:494:55:
called `Result::unwrap()` on an `Err` value: Compilation(Verifier(VerifierErrors([VerifierError { location: inst10, context: Some("v11 = icmp.i32 eq v9, v10  ; v10 = 0"), message: "arg 1 (v10) has type i64, expected i32" }])))
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: core::result::Result<T,E>::unwrap
             at /build/rustc-1.82.0-src/library/core/src/result.rs:1102:23
   4: roto::codegen::ModuleBuilder::define_function
             at /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/codegen/mod.rs:494:9
   5: roto::codegen::codegen
             at /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/codegen/mod.rs:353:9
   6: roto::pipeline::Lowered::codegen
             at /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/pipeline.rs:306:22
   7: roto::file_tree::FileTree::compile
             at /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/file_tree.rs:53:24
   8: roto_demo::main
             at ./src/main.rs:25:24
   9: core::ops::function::FnOnce::call_once
             at /build/rustc-1.82.0-src/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I have absolutely no clue what this means. So I figured: okay, I'll write an `fn path_eq(request: &Request, needle: Arc<str>) -> bool` function! Unfortunately, this doesn't compile:

```
❯ cargo build -q                                                                                                                           
error[E0277]: the trait bound `for<'a> Arc<str>: From<&'a IrValue>` is not satisfied
   --> src/main.rs:25:5
    |
25  |     #[roto_method(runtime, Request)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `for<'a> From<&'a IrValue>` is not implemented for `Arc<str>`, which is required by `DocumentedFunc<extern "C" fn(_, _, _)>: roto::runtime::func::Func<_, _>`
    |
    = help: the following other types implement trait `From<T>`:
              `Arc<str>` implements `From<&str>`
              `Arc<str>` implements `From<String>`
    = note: required for `&'a IrValue` to implement `for<'a> Into<Arc<str>>`
    = note: required for `Arc<str>` to implement `for<'a> TryFrom<&'a IrValue>`
    = note: required for `extern "C" fn(*mut bool, *const Request, Arc<str>)` to implement `roto::runtime::func::Func<(*mut bool, *const Request, Arc<str>), ()>`
    = note: 1 redundant requirement hidden
    = note: required for `DocumentedFunc<extern "C" fn(*mut bool, *const Request, Arc<str>)>` to implement `roto::runtime::func::Func<(*mut bool, *const Request, Arc<str>), ()>`
note: required by a bound in `roto::Runtime::register_method`
   --> /home/algernon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/roto-0.5.0/src/runtime/mod.rs:406:17
[...snip...]
```

`&str`, `String` don't work, and neither does `Arc<String>`.

I took another step back: what if I passed a String to a Roto function?

```rust
use roto::{src, Runtime, Verdict};
use std::sync::Arc;

fn main() {
    let s = src!(
        r#"
        filtermap main(path: String) {
          if path == "/" { accept } else { reject }
        }
        "#
    );
    let runtime = Runtime::new();

    let mut compiled = s.compile(runtime).unwrap();
    let func = compiled
        .get_function::<(), (Arc<str>,), Verdict<(), ()>>("main")
        .unwrap();

    let res = func.call(&mut (), "/".into());

    println!("{res:#?}");
}
```

Unfortunately, while this compiles, the Roto program compilation fails the same way as my first try.

And this is as far as I've got. If I'm doing something wrong, any pointers to how to do this kind of thing correctly would be highly appreciated.

--- a/src/codegen/tests.rs
+++ b/src/codegen/tests.rs
@@ -2090,3 +2090,43 @@ fn name_collision() {
         eprint!(".");
     }
 }
+
+#[test]
+fn str_equals() {
+    let s = src!(
+        r#"
+        function is_slash(s: String) -> bool {
+            s == "/"
+        }
+        "#
+    );
+    let runtime = Runtime::new();
+
+    let mut compiled = s.compile(runtime).unwrap();
+    let func = compiled
+        .get_function::<(), (Arc<str>,), bool>("is_slash")
+        .unwrap();
+
+    assert!(func.call(&mut (), "/".into()));
+    assert!(!func.call(&mut (), "foo".into()));
+}
+
+#[test]
+fn str_not_equals() {
+    let s = src!(
+        r#"
+        function is_not_slash(s: String) -> bool {
+            s != "/"
+        }
+        "#
+    );
+    let runtime = Runtime::new();
+
+    let mut compiled = s.compile(runtime).unwrap();
+    let func = compiled
+        .get_function::<(), (Arc<str>,), bool>("is_not_slash")
+        .unwrap();
+
+    assert!(func.call(&mut (), "foo".into()));
+    assert!(!func.call(&mut (), "/".into()));
+}

--- a/src/lower/mod.rs
+++ b/src/lower/mod.rs
@@ -734,7 +734,7 @@ impl<'r> Lowerer<'r> {
                     return Some(place.into());
                 }
 
-                if op == ast::BinOp::Eq && ty == Type::ip_addr() {
+                if op == ast::BinOp::Ne && ty == Type::ip_addr() {
                     let ip_addr_type_id = TypeId::of::<IpAddr>();
                     let runtime_func = self.find_runtime_function(
                         runtime::FunctionKind::Method(ip_addr_type_id),
@@ -750,6 +750,51 @@ impl<'r> Lowerer<'r> {
                         )
                         .unwrap();
 
+                    self.add(Instruction::Not {
+                        to: place.clone(),
+                        val: out,
+                    });
+                    return Some(place.into());
+                }
+
+                if op == ast::BinOp::Eq && ty == Type::string() {
+                    let str_type_id = TypeId::of::<Arc<str>>();
+                    let runtime_func = self.find_runtime_function(
+                        runtime::FunctionKind::Method(str_type_id),
+                        "eq",
+                    );
+
+                    let out = self
+                        .call_runtime_function(
+                            runtime_func.get_ref(),
+                            [left, right],
+                            &[Type::string(), Type::string()],
+                            &Type::bool(),
+                        )
+                        .unwrap();
+                    self.add(Instruction::Assign {
+                        to: place.clone(),
+                        val: out,
+                        ty: IrType::Bool,
+                    });
+                    return Some(place.into());
+                }
+
+                if op == ast::BinOp::Ne && ty == Type::string() {
+                    let str_type_id = TypeId::of::<Arc<str>>();
+                    let runtime_func = self.find_runtime_function(
+                        runtime::FunctionKind::Method(str_type_id),
+                        "eq",
+                    );
+
+                    let out = self
+                        .call_runtime_function(
+                            runtime_func.get_ref(),
+                            [left, right],
+                            &[Type::string(), Type::string()],
+                            &Type::bool(),
+                        )
+                        .unwrap();
                     self.add(Instruction::Not {
                         to: place.clone(),
                         val: out,

--- a/src/runtime/mod.rs
+++ b/src/runtime/mod.rs
@@ -1069,6 +1069,12 @@ impl Runtime {
             s.repeat(n as usize).into()
         }
 
+        /// Check for string equality
+        #[roto_method(rt, Arc<str>)]
+        fn eq(s: &Arc<str>, other: &Arc<str>) -> bool {
+            s == other
+        }
+
         rt
     }
 

